data-prepper-plugins/key-value-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java

```diff
@@ -179,79 +230,219 @@ private void validateKeySets(final Set<String> includeSet, final Set<String> exc
+        Map<Character, Character> bracketMap = initBracketMap();
+        int pairStart = 0;
+
+        ArrayList<String> pairs = new ArrayList<String>();
+        ObjectNode root = mapper.createObjectNode();
+
+        for (int i = 0; i < input.length(); i++) {
+            if (bracketMap.containsKey(input.charAt(i))) {
+                bracketStack.push(input.charAt(i));
+            }
 
-                if(value != null
-                        && value instanceof String
-                        && keyValueProcessorConfig.getDeleteValueRegex() != null
-                        && !Objects.equals(keyValueProcessorConfig.getDeleteValueRegex(), "")) {
-                    value = ((String)value).replaceAll(keyValueProcessorConfig.getDeleteValueRegex(), "");
+            if (bracketMap.containsValue(input.charAt(i)) && !bracketStack.isEmpty()) {
+                if (bracketMap.get(bracketStack.peek()) == input.charAt(i)) {
+                    bracketStack.pop();
                 }
+            }
+
+            if (bracketStack.isEmpty() && input.charAt(i) == fieldDelimiterPattern.toString().charAt(0)) {
+                String pair = input.substring(pairStart, i);
+                pairs.add(pair);
+                pairStart = i + fieldDelimiterPattern.toString().length();
+            }
+        }
+
+        pairs.add(input.substring(pairStart));
+
+        for (final String pair : pairs) {
+            int keyStart = 0;
+            int keyEnd = -1;
+            int valueStart = -1;
+            int valueEnd = -1;
+            String keyString = "";
+            String valueString;
 
-                if (keyValueProcessorConfig.getWhitespace().equals(whitespaceStrict)) {
-                    String[] whitespace_arr = trimWhitespace(key, value);
-                    key = whitespace_arr[0];
-                    value = whitespace_arr[1];
+            bracketStack.clear();
+
+            for (int i = 0; i < pair.length(); i++) {
+                if (bracketStack.isEmpty() && pair.charAt(i) == keyValueDelimiterPattern.toString().charAt(0)) {
+                    keyString = pair.substring(keyStart, i).stripTrailing();
+                    valueStart = i + keyValueDelimiterPattern.toString().length();
+                    while(pair.charAt(valueStart) == ' ') {
+                        valueStart++;
+                    }
+                    break;
                 }
+            }
 
-                if (keyValueProcessorConfig.getTransformKey() != null
-                        && !keyValueProcessorConfig.getTransformKey().isEmpty()) {
-                    key = transformKey(key);
+            if (keyString.isBlank()) {
+                keyString = pair;
+                LOG.debug("Unsuccessful match: '{}'", keyString);
+                valueString = keyValueProcessorConfig.getNonMatchValue().toString().stripLeading();
+            } else if (bracketMap.containsKey(pair.charAt(valueStart))) {
```

Comment:
I am not sure if I understand the logic below this if condition. Let's discuss offline.

Reply:
Discussed offline. It could be simplified into removing outermost bracket and pass into the next recursive call.